modules/dex-aggregator/src/lib.rs

```diff
@@ -0,0 +1,440 @@
+							if new_balance > optimal_balance {
+								optimal_balance = new_balance;
+								optimal_path = vec![pool.swap()];
+							}
+						}
+					}
+				}
+			} else if i == 1 {
+				let mut possible_paths = Vec::new();
+				for pool in all_pools.clone() {
+					if pair.first() == pool.1.first() {
+						possible_paths.push(pool);
+					} else if pair.first() == pool.1.second() {
+						possible_paths.push(pool.swap())
+					}
+					for pool1 in possible_paths.clone() {
+						for pool2 in all_pools.clone() {
+							if pair.second() == pool2.1.second() && pool1.1.second() == pool2.1.first() {
+								if let Some(new_balance) = Self::get_target_amount(vec![pool1, pool2], supply_amount) {
+									if new_balance > optimal_balance {
+										optimal_balance = new_balance;
+										optimal_path = vec![pool1, pool2];
+									}
+								}
+							} else if pair.second() == pool2.1.first() && pool1.1.second() == pool2.1.second() {
+								if let Some(new_balance) =
+									Self::get_target_amount(vec![pool1, pool2.swap()], supply_amount)
+								{
+									if new_balance > optimal_balance {
+										optimal_balance = new_balance;
+										optimal_path = vec![pool1, pool2.swap()];
+									}
+								}
+							}
+						}
+					}
+				}
+			}
+			i += 1;
+		}
+
+		if optimal_path.is_empty() {
+			None
+		} else {
+			Some((optimal_path, optimal_balance))
+		}
+	}
+
+	/// Returns tuple of optimal path with expected supply amount. Returns None if trade is not
+	/// possible. TODO: see above to do in optimal_path_with_exact_supply
+	fn optimal_path_with_exact_target(
+		pair: TradingDirection,
+		target_amount: Balance,
+	) -> Option<(Vec<AvailablePool>, Balance)> {
+		let mut i: u32 = 0;
+		let all_pools = Self::all_active_pairs();
+		let mut optimal_path: Vec<AvailablePool> = Vec::new();
+		let mut optimal_balance: Balance = u128::MAX;
+
+		while i < T::AggregatorTradingPathLimit::get() {
+			if i == 0 {
+				for pool in all_pools.clone() {
+					if pair == pool.1 {
+						if let Some(new_balance) = Self::get_supply_amount(vec![pool], target_amount) {
+							if new_balance < optimal_balance {
+								optimal_balance = new_balance;
+								optimal_path = vec![pool];
+							}
+						}
+					}
+
+					if pair == pool.1.swap() {
+						if let Some(new_balance) = Self::get_supply_amount(vec![pool.swap()], target_amount) {
+							if new_balance < optimal_balance {
+								optimal_balance = new_balance;
+								optimal_path = vec![pool.swap()];
+							}
+						}
+					}
+				}
+			} else if i == 1 {
+				let mut possible_paths = Vec::new();
+				for pool in all_pools.clone() {
+					if pair.first() == pool.1.first() {
+						possible_paths.push(pool);
+					} else if pair.first() == pool.1.second() {
+						possible_paths.push(pool.swap())
+					}
+					for pool1 in possible_paths.clone() {
+						for pool2 in all_pools.clone() {
+							if pair.second() == pool2.1.second() && pool1.1.second() == pool2.1.first() {
```

Comment:
We already know that the exchange path is `supply_token -> target_token`, so we can find the `middle_token` which can exchange to `supply_token` and `target_token`, then generate the path `supply_token -> middle_token -> target_token`

Reply:
see above comments